galsim/phase_psf.py

```diff
@@ -620,6 +620,23 @@ def __getstate__(self):
             d.pop('_illuminated', None)
         return d
 
+    def samplePupil(self, photons, rng):
+        """Set the pupil_u and pupil_v values in the PhotonArray by sampling the current aperture.
+        """
+        from .random import UniformDeviate
+
+        n_photons = len(photons)
+        u = self.u_illuminated
+        v = self.v_illuminated
+        ud = UniformDeviate(rng)
+        pick = np.empty((n_photons,), dtype=float)
+        ud.generate(pick)
+        pick *= len(u)
+        pick = pick.astype(int)
+        photons.pupil_u = u[pick]
+        photons.pupil_v = v[pick]
```

Comment:
I thought you might add to this to get a more continuous distribution of u,v values.  Maybe add a random +-0.5 pixel in each direction?
```
duv = pick   # reuse this array, but renamed to not be confusing.
ud.generate(duv)
duv -= 0.5
duv *= self.pupil_plane_scale
photons.pupil_u += duv[:len(photons.pupil_u)]
ud.generate(duv)
duv -= 0.5
duv *= self.pupil_plane_scale
photons.pupil_v += duv[:len(photons.pupil_v)]
```
And all of this boilerplate is making me want to do #1067...   :)

Reply:
> Maybe add a random +-0.5 pixel in each direction?

Sure.  I guess that's a bit wrong at the edges, but probably not more wrong than the current strategy.

Reply:
> And all of this boilerplate is making me want to do https://github.com/GalSim-developers/GalSim/issues/1067... :)

Not sure, but I think the current version of this might just be to subclass [`numpy.random.BitGenerator`](https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.BitGenerator.html#numpy.random.BitGenerator) using `BaseDeviate` to generate the bits.

Reply:
"just" is not as easy as the docs imply.  Here is the longer explanation of how to subclass this: https://numpy.org/doc/stable/reference/random/extending.html